File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/runtime/iterators/ResetIterator.java
package org.opencds.cqf.cql.engine.runtime.iterators;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by Bryn on 8/11/2019.
 */
public class ResetIterator<E> implements Iterator<E> {
    private Iterator<E> source;
    private ArrayList<E> data = new ArrayList<E>();
    int dataIndex = -1;
    private boolean dataCached = false;

    public ResetIterator(Iterator<E> source) {
        this.source = source;
        data = new ArrayList<E>();
    }

    @Override
    public boolean hasNext() {
        if (!dataCached) {
            return source.hasNext();
        }

        return dataIndex < data.size() - 1 && data.size() > 0;
    }

    @Override
    public E next() {
        if (!dataCached) {
            E element = source.next();
            data.add(element);
            return element;
        }

        dataIndex++;
        return data.get(dataIndex);
    }

    public void reset() {
        // Fill any remaining data
        while (source.hasNext()) {
            data.add(source.next());
        }
        dataCached = true;
        dataIndex = -1;
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/runtime/iterators/ResetIterator.kt
package org.opencds.cqf.cql.engine.runtime.iterators

/** Created by Bryn on 8/11/2019. */
class ResetIterator<E>(private val source: Iterator<E?>) : Iterator<E?> {
    private var data = ArrayList<E?>()
    var dataIndex: Int = -1
    private var dataCached = false

    init {
        data = ArrayList<E?>()
    }

    override fun hasNext(): Boolean {
        if (!dataCached) {
            return source.hasNext()
        }

        return dataIndex < data.size - 1 && data.isNotEmpty()
    }

    override fun next(): E? {
        if (!dataCached) {
            val element = source.next()
            data.add(element)
            return element
        }

        dataIndex++
        return data[dataIndex]
    }

    fun reset() {
        // Fill any remaining data
        while (source.hasNext()) {
            data.add(source.next())
        }
        dataCached = true
        dataIndex = -1
    }
}
